extensions/amp-social-share/0.1/amp-social-share.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getSocialConfig} from './amp-social-share-config';
+
+const DEFAULT_WIDTH = 60;
+const DEFAULT_HEIGHT = 44;
+
+const CONTAINER_TAG = "span";
+const LINK_TAG = "a";
+
+class AmpSocialShare extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  buildCallback() {
+    this.config = this.getElementConfig(this.element);
+    AMP.assert(this.config, 'The <script> config is required for ' +
+      '<amp-social-share> %s');
+
+    this.type = AMP.assert(this.element.getAttribute('type'),
+        'The type attribute is required for <amp-social-share> %s',
+        this.element);
+    this.typeConfig = getSocialConfig(this.type);
+
+    this.container = this.renderShare(this.config);
+
+    this.width = this.element.getAttribute('width') || DEFAULT_WIDTH;
+    this.height = this.element.getAttribute('height') || DEFAULT_HEIGHT;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.element.appendChild(this.container);
+    return Promise.resolve();
+  }
+
+  renderShare(config) {
+    const urlParams = {};
+
+    for (const param in this.typeConfig["params"]) {
+      const paramConf = this.typeConfig["params"][param];
+      AMP.assert(!paramConf["required"] || param in config,
+        param + " is a required configuration item for " + this.type);
+
+      if (param in config) {
+        const value = config[param];
+        if ("maxlength" in paramConf) {
+          const maxlength = paramConf["maxlength"];
+          AMP.assert(value.length < maxlength,
+            param + " cannot exceed " + maxlength);
+        }
+        urlParams[paramConf["param"]] = value;
+      } else if (paramConf["type"] == "fixed") {
+        urlParams[paramConf["param"]] = paramConf["value"];
+      }
+    }
+    let container = this.getFirstSubElement(this.element, CONTAINER_TAG);
+    let link = this.getFirstSubElement(this.element, LINK_TAG);
+    if (container === null) {
+      container = this.getWin().document.createElement(CONTAINER_TAG);
+      container.setAttribute("class", "-amp-social-share -amp-social-share-" +
+        this.type);
+    }
+    if (link === null) {
+      link = this.getWin().document.createElement(LINK_TAG);
+      link.innerHTML = this.typeConfig["text"];
```

Comment:
It was used to allow the amp-social-share-config.js to specify html for inside the anchor. I converted it to a createTextNode before, but that prohibited being able to set &nbsp; (as a no anchor text). It's not content being injected from an untrusted source, I'm just trying to offer the maximum flexibility.

I'm totally open to implementation suggestions.